File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

ageOn :: Planet -> Float -> Float
ageOn planet seconds =
    case planet of
        Mercury -> age / 0.2408467
        Venus -> age /  0.61519726
        Earth -> age / 1.0
        Mars -> age / 1.8808158
        Jupiter -> age / 11.862615
        Saturn -> age / 29.447498
        Uranus -> age / 84.016846
        Neptune -> age / 164.79132
    where age = seconds / 365.25 / 24 / 60 / 60

File: c/space-age/space_age.c
#include "space_age.h"

#define EARTH_SEC 31557600.0

static float periods[] = {
      0.2408467,
      0.61519726,
      1.0,
      1.8808158,
     11.862615,
     29.447498,
     84.016846,
    164.79132
};

float age(planet_t planet, int64_t seconds) {
    if (planet < MERCURY || planet > NEPTUNE) return -1;

    return seconds / EARTH_SEC / periods[planet];
}
